src/value.rs

```diff
@@ -1,20 +1,43 @@
-use std::ops::{Add, Div, Mul, Sub};
+use derivative::Derivative;
+use std::ops::{Add, Div, Mul, Neg, Sub};
 
 /*
  * A wrapper around a numerical value, which
  * performs a derivative computation in the tangent mode.
  */
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Derivative)]
+#[derivative(Debug, Clone, Copy, Default)]
 pub struct Value {
     pub value: f32,
+    #[derivative(Default(value = "0.0"))]
     pub der: f32,
 }
 
 impl Value {
     pub fn new(value: f32, der: f32) -> Self {
         Value { value, der }
     }
+
+    pub fn pow(self, exp: f32) -> Self {
+        let value = self.value.powf(exp);
+        let der = exp * self.value.powf(exp - 1.0) * self.der;
+        Value { value, der }
+    }
+
+    pub fn sqrt(self) -> Self {
+        let value = self.value.sqrt();
+        let der = 0.5 * self.value.powf(-0.5) * self.der;
+        Value { value, der }
+    }
+
+    pub fn relu(self) -> Self {
```

Comment:
relu is very critical for the NNs, but I wonder if the non-differentiability at 0 needs to be somehow taken into account.

Reply:
I thought about it as well but I could not think of an edge case where this would become important. But I can do a bit more research and cross check with some more sophisticated libraries.

Reply:
This is the implementation of micrograd:

```python
    def relu(self):
        out = Value(0 if self.data < 0 else self.data, (self,), 'ReLU')

        def _backward():
            self.grad += (out.data > 0) * out.grad
        out._backward = _backward

        return out
```

Also [Wikipedia](https://en.wikipedia.org/wiki/Rectifier_(neural_networks)) mentions:
> ***Potential problems:***
> *- Non-differentiable at zero; however, it is differentiable anywhere else, and the value of the derivative at zero can be arbitrarily chosen to be 0 or 1.*